larpandoracontent/LArHelpers/LArHierarchyHelper.cc

```diff
@@ -0,0 +1,1171 @@
+
+void LArHierarchyHelper::MCHierarchy::Node::FillHierarchy(const MCParticle *pRoot, const bool foldToLeadingShowers)
+{
+    MCParticleList allParticles{pRoot};
+    int pdg{std::abs(pRoot->GetParticleId())};
+    const bool isShower{pdg == E_MINUS || pdg == PHOTON};
+    const bool isNeutron{pdg == NEUTRON};
+    if (foldToLeadingShowers && (isShower || (isNeutron && !m_hierarchy.m_recoCriteria.m_removeNeutrons)))
+        LArMCParticleHelper::GetAllDescendentMCParticles(pRoot, allParticles);
+    else if (m_hierarchy.m_recoCriteria.m_removeNeutrons && isNeutron)
+        return;
+    CaloHitList allHits;
+    for (const MCParticle *pMC : allParticles)
+    {
+        // ATTN - Not all MC particles will have hits
+        if (m_hierarchy.m_mcToHitsMap.find(pMC) != m_hierarchy.m_mcToHitsMap.end())
+        {
+            const CaloHitList &caloHits(m_hierarchy.m_mcToHitsMap.at(pMC));
+            allHits.insert(allHits.begin(), caloHits.begin(), caloHits.end());
+        }
+    }
+    if (!allParticles.empty())
+    {
+        Node *pNode{new Node(m_hierarchy, allParticles, allHits)};
+        m_children.emplace_back(pNode);
+        if (!foldToLeadingShowers || (foldToLeadingShowers && !(isShower || isNeutron)))
+        {   // Find the children of this particle and recursively add them to the hierarchy
+            const MCParticleList &children{pRoot->GetDaughterList()};
+            for (const MCParticle *pChild : children)
+                pNode->FillHierarchy(pChild, foldToLeadingShowers);
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void LArHierarchyHelper::MCHierarchy::Node::FillFlat(const MCParticle *pRoot)
+{
+    MCParticleList allParticles{pRoot};
+    if (!m_hierarchy.m_recoCriteria.m_removeNeutrons)
+    {
+        LArMCParticleHelper::GetAllDescendentMCParticles(pRoot, allParticles);
+    }
+    else
+    {
+        MCParticleList neutrons;
+        LArMCParticleHelper::GetAllDescendentMCParticles(pRoot, allParticles, allParticles, neutrons);
+    }
+    CaloHitList allHits;
+    for (const MCParticle *pMC : allParticles)
+    {
+        // ATTN - Not all MC particles will have hits
+        if (m_hierarchy.m_mcToHitsMap.find(pMC) != m_hierarchy.m_mcToHitsMap.end())
+        {
+            const CaloHitList &caloHits(m_hierarchy.m_mcToHitsMap.at(pMC));
+            allHits.insert(allHits.begin(), caloHits.begin(), caloHits.end());
+        }
+    }
+    if (!allParticles.empty())
+    {
+        Node *pNode{new Node(m_hierarchy, allParticles, allHits)};
+        m_children.emplace_back(pNode);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool LArHierarchyHelper::MCHierarchy::Node::IsReconstructable() const
+{
+    const bool enoughHits{m_caloHits.size() >= m_hierarchy.m_recoCriteria.m_minHits};
```

Comment:
Couldn't you return false here if enoughHits is false and save doing the rest of the function?